### PR TITLE
Simulator: Add basic PhysicsRobot class

### LINKED ISSUE #883: SimpleSimulator: Create a PhysicsRobot abstraction
### Description of the task

See #768 for more information on the Simple Simulator.

Create a PhysicsRobot class to make it easy to create Robots in the Box2D physics world and convert the data back to our own Robot class.
* The Box2D object created will be responsible for handling collisions and physics
* The PhysicsRobot should expose functions that would allow control over the basic hardware capabilities, such as autokick, breakbeam, and implementing primitives.

### Acceptance criteria

- [ ] Create a PhysicsRobot class
- [ ] Unit tests

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

Adds a `PhysicsRobot` class to represent a `Robot` in simulation. This PR just sets up the class and the basic physics bodies, it does not implement robot capabilities like kicking.

### Testing Done

Unit tests

### Resolved Issues

`partially resolves #883`
(formatting this so the ticket won't auto-close)

### Length Justification

Lots of unit tests, puts it just over.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by MathewMacDougall** [comment]
@garethellis0 please feel free to review this before others (rather than waiting if you see it first) since this blocks my next steps.

File: src/software/backend/simulation/physics/physics_robot.cpp [hunk 1]
```diff
@@ -0,0 +1,120 @@
+#include "software/backend/simulation/physics/physics_robot.h"
+
+#include "shared/constants.h"
+#include "software/backend/simulation/physics/box2d_util.h"
+
+PhysicsRobot::PhysicsRobot(std::shared_ptr<b2World> world, const Robot& robot)
+    : robot_id(robot.id())
+{
+    // All the BodyDef must be defined before the body is created.
```
Comment (garethellis0) [inline src/software/backend/simulation/physics/physics_robot.cpp]: what's a `BodyDef`?

Comment (garethellis0) [reply src/software/backend/simulation/physics/physics_robot.cpp]: Also I'm a bit confused about the order stuff is happening here. How do we know this is "before the body is created"?

File: src/software/backend/simulation/physics/physics_robot.h [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+     * @param robot The Robot being created in the Box2D world
+     */
+    void createRobotPhysicsBody(std::shared_ptr<b2World> world, const Robot& robot);
+
+    /**
+     * A helper function that sets up the robot body physics object
+     *
+     * @param robot The Robot being created in the Box2D world
+     */
+    void setupRobotPhysicsBody(const Robot& robot);
```
Comment (garethellis0) [inline src/software/backend/simulation/physics/physics_robot.h:66]: Just looking at the jdocs for `setupRobotPhysicsBody` and  `createRobotPhysicsBody`, bit confused on what the difference between them is?

File: src/software/backend/simulation/physics/physics_robot_test.cpp [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+{
+    b2Vec2 gravity(0, 0);
+    auto world = std::make_shared<b2World>(gravity);
+
+    Robot robot_parameter(0, Point(0, 0), Vector(0, 0), Angle::zero(),
+                          AngularVelocity::zero(), Timestamp::fromSeconds(0));
+    PhysicsRobot physics_robot(world, robot_parameter);
+    auto robot = physics_robot.getRobotWithTimestamp(Timestamp::fromSeconds(3.3));
+
+    // TODO: CHECK ROBOT STATE
```
Comment (garethellis0) [inline src/software/backend/simulation/physics/physics_robot_test.cpp]: TODO?

**Review by garethellis0 — CHANGES_REQUESTED** [review]

File: src/shared/constants.h [hunk 4]
```diff
@@ -12,6 +12,8 @@ const double BALL_MAX_SPEED_METERS_PER_SECOND = 6.5;
 const double ROBOT_MAX_RADIUS_METERS = 0.09;
 // The distance from the center of the robot to the front face (the flat part), in meters
 const double DIST_TO_FRONT_OF_ROBOT_METERS = 0.07;
+// The total width of the entire flat face on the front of the robot
+const double FRONT_OF_ROBOT_WIDTH = 0.11;
```
Comment (EvanMorcom) [inline src/shared/constants.h]: Mega nit:

Maybe add units here

File: src/software/backend/simulation/physics/physics_robot.cpp [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        robot_body_world->DestroyBody(robot_body);
+    }
+}
+
+void PhysicsRobot::createRobotPhysicsBody(std::shared_ptr<b2World> world,
+                                          const Robot& robot)
+{
+    robot_body_def.type = b2_dynamicBody;
+    robot_body_def.position.Set(robot.position().x(), robot.position().y());
+    robot_body_def.linearVelocity.Set(robot.position().x(), robot.position().y());
```
Comment (EvanMorcom) [inline src/software/backend/simulation/physics/physics_robot.cpp]: You are using position here instead of velocity.
Typo?

File: src/software/backend/simulation/physics/physics_robot.cpp [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+    robot_body->CreateFixture(&robot_fixture_def);
+}
+
+void PhysicsRobot::setupChickerPhysicsBody(const Robot& robot)
+{
+    // Create a very thin rectangle where the dribbler and chicker are on the
+    // flat face of the robot
+    robot_chicker_shape.SetAsBox(
+        0.001, DRIBBLER_WIDTH / 2.0,
```
Comment (EvanMorcom) [inline src/software/backend/simulation/physics/physics_robot.cpp]: Nit:
Maybe create a constant for the thickness here

**Review by EvanMorcom — CHANGES_REQUESTED** [review]
:fire:

File: src/software/backend/simulation/physics/physics_robot.cpp [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    {
+        Angle angle_to_vertex = global_angle_to_edge_of_flat_face + (i * angle_increment);
+        // Shapes are defined relative to the body they are added to, so we do everything
+        // relative to the centre of the robot body, which is treated as (0, 0)
+        Vector vector_from_body_to_vertex =
+            Vector::createFromAngle(angle_to_vertex).normalize(ROBOT_MAX_RADIUS_METERS);
+        Point vertex           = Point(0, 0) + vector_from_body_to_vertex;
+        robot_body_vertices[i] = createVec2(vertex);
+    }
+
```
Comment (jonathanlew) [inline src/software/backend/simulation/physics/physics_robot.cpp]: Would look cleaner if we put this into a helper function such as `getRobotBodyVertices`

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Nice work! Just a little suggestion

File: src/software/backend/simulation/physics/physics_robot.cpp (hunk 1 above)
Comment (MathewMacDougall) [reply src/software/backend/simulation/physics/physics_robot.cpp]: Changed this comment as it wasn't appropriate here. Clarified with a new comment.

tl;dr a `BodyDef` is a struct that defines properties of a `b2Body` in the physics engine.

File: src/software/backend/simulation/physics/physics_robot.h (hunk 2 above)
Comment (MathewMacDougall) [reply src/software/backend/simulation/physics/physics_robot.h:66]: Clarified the comments. lmk if that's better.

File: src/software/backend/simulation/physics/physics_robot_test.cpp (hunk 3 above)
Comment (MathewMacDougall) [reply src/software/backend/simulation/physics/physics_robot_test.cpp]: Waiting on #1037 . Removed the `TODO` for now as it's non-blocking (for now)

File: src/shared/constants.h (hunk 4 above)
Comment (MathewMacDougall) [reply src/shared/constants.h]: :+1: fixed

File: src/software/backend/simulation/physics/physics_robot.cpp (hunk 5 above)
Comment (MathewMacDougall) [reply src/software/backend/simulation/physics/physics_robot.cpp]: massive typo. Fixed

File: src/software/backend/simulation/physics/physics_robot.cpp (hunk 6 above)
Comment (MathewMacDougall) [reply src/software/backend/simulation/physics/physics_robot.cpp]: I don't think thickness is important because this is just for simulation. We have no use for the thickness of the dribbler IRL, and it's harder to define.

What do you think?

File: src/software/backend/simulation/physics/physics_robot.cpp (hunk 7 above)
Comment (MathewMacDougall) [reply src/software/backend/simulation/physics/physics_robot.cpp]: Agreed. Refactored into new function.

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

File: src/software/backend/simulation/physics/physics_robot.cpp (hunk 6 above)
Comment (EvanMorcom) [reply src/software/backend/simulation/physics/physics_robot.cpp]: Sounds good to me. 

I was looking at it as more of a legibility thing

**Review by EvanMorcom — APPROVED** [review]
LGTM :100:

**Review by jonathanlew — APPROVED** [review]
LGTM!
